File: dash/frontend/src/types/onboarding.ts
export interface OnboardingState {
  current_step: string;
  completed_steps: string[];
  site_id: string | null;
  scope_id: string | null;
  first_job_id: string | null;
  demo_used: boolean;
  dismissed: boolean;
  completed_at: string | null;
}

export interface CompleteStepPayload {
  step: string;
  site_id?: string;
  scope_id?: string;
  first_job_id?: string;
  demo_used?: boolean;
}

export interface RecoveryCodes {
  codes: string[];
  generated_at: string;
}

export interface NetworkCandidates {
  candidates: string[];
  source: string;
  note: string;
}

export interface ScopePreview {
  cidr: string;
  host_estimate: number;
  is_private: boolean;
  warnings: string[];
  requires_confirmation: boolean;
}

export interface ScanPreset {
  key: string;
  name: string;
  mode: string;
  description: string;
  checks: string[];
  intrusive: boolean;
  active_web: boolean;
  uses_credentials: boolean;
  resource_class: string;
  duration_class: string;
}

export interface ScanSummary {
  preset: string;
  preset_name: string;
  targets: string[];
  host_estimate: number;
  checks: string[];
  intrusive: boolean;
  active_web: boolean;
  uses_credentials: boolean;
  resource_class: string;
  duration_class: string;
  demo: boolean;
  data_retention: string;
}

export interface DemoTarget {
  cidr: string;
  note: string;
}

export interface ComponentHealth {
  application: string;
  database: string;
  local_scout: string;
  scanner_capabilities: string;
  feeds: string;
}

export interface ProbeSummary {
  id: string;
  name: string;
  status: string;
  site_id: string;
  pentest_enabled?: boolean;
  credentialed_scans_enabled?: boolean;
  has_encryption_key?: boolean;
}

export interface ProbeDetail {
  id: string;
  site_id: string;
  name: string;
  description: string | null;
  status: string;
  online: boolean;
  certificate_fingerprint: string;
  agent_version: string | null;
  operating_system: string | null;
  architecture: string | null;
  hostname: string | null;
  primary_ip: string | null;
  pentest_enabled: boolean;
  credentialed_scans_enabled: boolean;
  has_encryption_key: boolean;
  last_seen_at: string | null;
  enrolled_at: string | null;
  approved_at: string | null;
  created_at: string;
}

export interface JobSummary {
  id: string;
  status: string;
  mode: string;
}

export interface ProfilePlanQuestion {
  key: string;
  label: string;
  kind: 'boolean' | 'number' | 'text' | 'select';
  options: string[];
  required: boolean;
}

export interface ProfileRecommendation {
  capability: string;
  status: 'available' | 'planned';
  reason: string;
  route: string | null;
}

export interface ProfilePlan {
  experience_profile: string;
  questions: ProfilePlanQuestion[];
  answers: Record<string, unknown>;
  recommendations: ProfileRecommendation[];
  updated_at: string | null;
}

File: dash/backend/app/schemas/onboarding.py
"""Guided first-run (onboarding) schemas (Phase 19)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class OnboardingStateRead(BaseModel):
    """Resumable wizard state."""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    current_step: str
    completed_steps: list[str] = Field(validation_alias="completed_steps_json")
    site_id: uuid.UUID | None
    scope_id: uuid.UUID | None
    first_job_id: uuid.UUID | None
    demo_used: bool
    dismissed: bool
    completed_at: datetime | None


class CompleteStepRequest(BaseModel):
    step: str
    site_id: uuid.UUID | None = None
    scope_id: uuid.UUID | None = None
    first_job_id: uuid.UUID | None = None
    demo_used: bool | None = None


class RecoveryCodesResponse(BaseModel):
    """Plaintext recovery codes — shown exactly once."""

    codes: list[str]
    generated_at: datetime


class NetworkCandidatesResponse(BaseModel):
    candidates: list[str]
    source: str
    note: str


class ScopePreviewRequest(BaseModel):
    cidr: str
    allow_public: bool = False


class ScopePreviewResponse(BaseModel):
    cidr: str
    host_estimate: int
    is_private: bool
    warnings: list[str]
    requires_confirmation: bool


class ScanPreset(BaseModel):
    key: str
    name: str
    mode: str
    description: str
    checks: list[str]
    intrusive: bool
    active_web: bool
    uses_credentials: bool
    resource_class: str
    duration_class: str


class ScanPresetsResponse(BaseModel):
    presets: list[ScanPreset]


class ScanSummaryRequest(BaseModel):
    preset: str = "standard"
    targets: list[str]
    demo: bool = False


class ScanSummaryResponse(BaseModel):
    preset: str
    preset_name: str
    targets: list[str]
    host_estimate: int
    checks: list[str]
    intrusive: bool
    active_web: bool
    uses_credentials: bool
    resource_class: str
    duration_class: str
    demo: bool
    data_retention: str


class DemoTargetResponse(BaseModel):
    cidr: str
    note: str


class ProfilePlanQuestion(BaseModel):
    key: str
    label: str
    kind: Literal["boolean", "number", "text", "select"]
    options: list[str] = Field(default_factory=list)
    required: bool = False


class ProfileRecommendation(BaseModel):
    capability: str
    status: Literal["available", "planned"]
    reason: str
    route: str | None = None


class ProfilePlanRead(BaseModel):
    experience_profile: str
    questions: list[ProfilePlanQuestion]
    answers: dict[str, Any]
    recommendations: list[ProfileRecommendation]
    updated_at: datetime | None = None


class ProfilePlanUpdate(BaseModel):
    answers: dict[str, Any] = Field(default_factory=dict)
